man/man1/testbsc.1
.TH TESTBSC 1NEMO "3 July 2019"
.SH NAME
testbsc \- testing Beam Smearing Correction algorithms
.SH SYNOPSIS
\fBtestbsc\fP [parameter=value]
.SH DESCRIPTION
\fBtestbsc\fP tests various Beam Smearing Correction schemes to a velocity field. Currently
only Begeman's correction is available, but expect to try out Leung's (2018) method
in an upcoming version.
.PP
Begeman's (1989, Appendix B, eq.(8)) recipe reads as follows:
.nf
     v = V + b^2/n * ( Vx*Nx + Vy*Ny + N * (Vxx + Vyy) / 2)
.fi
where the correction terms (\fBbeeg=t\fP below) have single and double
derivatives. (N,V) are the model density and velocity, (n,v) the
observed/smoothed values, in this case with a circular gaussian beam
with sigma (not FWHM) of \fBb\fP (see \fBbeam=\fP below) points per pixel.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP
\fBinD=\fP
Input Model Density image file []  
.TP
\fBinV=\fP
Input Model Velocity image file []  
.TP
\fBind=\fP
Input Observed (smoothed) Density image file []  
.TP
\fBinv=\fP
Input Observed (smoothed) Velocity image file []  
.TP
\fBout=\fP
Output file, depends on the values of \fBdiff=\fP and \fBbeeg=\fP.
.TP
\fBbeam=\fP
Override sigma_beam/pixel ratio. Recall FWHM = 2.355 * SIGMA. By default the beam
size will be in the header.
.TP
\fBdiff=f|t\fP
Compute a Difference map? (as opposed to Computed Observed). Normally the Observed velocity field is
computed from the Model and BSC terms, but turning this option on,
the input Observed is subtracted from the computed Observed. [f]
.TP
\fBbeeg=t|f\fP
Use the Begeman correction terms. This is of course the default, but by
turning this off, a simple different of V-v (Model - Observed) can be computed, as
can be seen in the EXAMPLES below. [t]
.SH SEE ALSO
ccdmath(1NEMO), rotcur(1NEMO), image(5NEMO)
.SH EXAMPLES
In this example we show off the interplay of the \fBdiff=\fP and \fBbeeg=\fP keywords:
.nf
  testbsc N.ccd V.ccd n.ccd v.ccd map1 diff=f beeg=f
  testbsc N.ccd V.ccd n.ccd v.ccd map2 diff=t beeg=f
  testbsc N.ccd V.ccd n.ccd v.ccd map3 diff=f beeg=t
  testbsc N.ccd V.ccd n.ccd v.ccd map4 diff=t beeg=t
.fi
Now the four maps will have the following meaning:
.nf
  map1:  V
  map2:  V - v
  map3:  V + bsc       (should be close to 'v')
  map4:  V + bsc - v   (should be close to 0)
.fi
.SH FILES
NEMO/src/image/rotcur
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
3-jul-2019	V0.4 Created/released	PJT
.fi


src/image/rotcur/testbsc.c
/* 
 * TESTBSC: test Beam Smearing Correction code
 *
 *      2-jul-2019    Begeman's appendix B
 *
 *
 * We will be using our (D,V,S) and (d,v,s) notation, not Begeman's (N,V,S) and (n,w,.)
 *
 */


#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <image.h>

string defv[] = {
        "inD=\n        Input Model Density image file",
        "inV=\n        Input Model Velocity image file",
        "ind=\n        Input Observed Density image file",
        "inv=\n        Input Observed Velocity image file",
	"out=\n        Output (TBD)",
	"beam=\n       Override beam/pixel ratio from header value",
	"diff=f\n      Difference Map? (as opposed to Observed)",
	"beeg=t\n      Use the Begeman correction terms to V",
	"VERSION=0.4\n 3-jul-2019 PJT",
	NULL,
};

string usage = "testing Beam Smearing Corrections";

#define CV01(x,y)  MapValue(iptr01,x,y)
#define CV02(x,y)  MapValue(iptr02,x,y)
#define CV11(x,y)  MapValue(iptr11,x,y)
#define CV12(x,y)  MapValue(iptr12,x,y)
#define CVO(x,y)   MapValue(optr,x,y)


void nemo_main()
{
    stream   instr01,instr02,instr11,instr12,outstr;
    int      nx, ny;
    int      i,j;
    imageptr iptr01=NULL, iptr02=NULL;      /* pointer to images */
    imageptr iptr11=NULL, iptr12=NULL;
    imageptr optr = NULL;
    imageptr iptr = NULL;
    real     d1, d2, d3, d4, dx, dy, b;
    real     v1, n1, v2, n2;
    bool     Qdiff = getbparam("diff");
    bool     Qbeeg = getbparam("beeg");

    if (hasvalue("inD")) {
      instr01 = stropen(getparam("inD"),"r");
      read_image( instr01, &iptr01);
      iptr = iptr01;
    }
    if (hasvalue("inV")) {
      instr02 = stropen(getparam("inV"),"r");
      read_image( instr02, &iptr02);
      iptr = iptr02;      
    }
    if (hasvalue("ind")) {
      instr11 = stropen(getparam("ind"),"r");
      read_image( instr11, &iptr11);
      iptr = iptr11;      
    }
    if (hasvalue("inv")) {
      instr12 = stropen(getparam("inv"),"r");
      read_image( instr12, &iptr12);
      iptr = iptr12;      
    }
    if (iptr == NULL) error("No input image(s) given");
    
    nx = Nx(iptr);	
    ny = Ny(iptr);
    dx = Dx(iptr);
    dy = Dy(iptr);
    /* @todo    should check if maps are same size */

    b = Beamy(iptr) / Dy(iptr) / 2.355;    // 2.355 = 2*sqrt(2*ln(2))
    if (hasvalue("beam"))
      b = getrparam("beam");

    dprintf(0,"Maps size %d x %d assumed,   beam/pixel = %g\n",nx,ny, b);

    outstr = stropen(getparam("out"), "w");
    create_cube(&optr,nx,ny,1);
    Dx(optr) = Dx(iptr);
    Dy(optr) = Dy(iptr);
    Xmin(optr) = Xmin(iptr);
    Ymin(optr) = Ymin(iptr);
    /* should do the others too */

    for (j=1; j<ny-1; j++) {
      for (i=1; i<nx-1; i++) {
	if (CV01(i,j) == 0) {
	  CVO(i,j) = 0;        // blank out where model is 0
	  continue;
	}
	// 2nd order terms (Laplacian)
	d1 = CV02(i-1,j) - CV02(i,j);
	d2 = CV02(i+1,j) - CV02(i,j); 
	d3 = CV02(i,j-1) - CV02(i,j);
	d4 = CV02(i,j+1) - CV02(i,j); 
	// linear terms
	n1 = 0.5*(CV01(i+1,j) - CV01(i-1,j));    // dN/dx
	v1 = 0.5*(CV02(i+1,j) - CV02(i-1,j));    // dV/dx
	n2 = 0.5*(CV01(i,j+1) - CV01(i,j-1));    // dN/dy
	v2 = 0.5*(CV02(i,j+1) - CV02(i,j-1));    // dV/dy
	//
	CVO(i,j)  = CV02(i,j);
	if (Qbeeg) {
	  CVO(i,j) += (v1*n1 + v2*n2) * b * b / CV11(i,j);                   // linear
	  CVO(i,j) += (d1+d2+d3+d4) * CV01(i,j) * b * b / ( 2 * CV11(i,j));  // 2nd order
	}
	if (Qdiff)
	  CVO(i,j) -= CV12(i,j);
      }
    }
    write_image(outstr, optr);
}
